### ISSUE TITLE
Handle disconnecting and reconnecting

### ISSUE THREAD

**Issue by mtbentley** [ajm188/mqtt-locust#12]
(This might fix the 104 errors)

**Comment by mtbentley** [comment]
Also, each publish opens a new connection to mqtt, which is neither optimal nor what might be expected in normal situations

**Comment by ajm188** [comment]
Each publish doesn't, but each "user" hatched does.

**Comment by mtbentley** [comment]
Yeah, I just realized that.  I still don't think mosquitto is designed to handle 1000s of users, but instead a lot of requests per second

**Comment by ajm188** [comment]
Have each publishing user take an argument that says how many publishes they should spam each time it's their turn?

**Comment by mtbentley** [comment]
That could work.  Any idea how locust would handle it?

**Comment by ajm188** [comment]
You could just have it as a kwarg on the task set. So something like

``` python
class MyBehavior(TaskSet):
    @task(1)
     def spammy_publish(self, *args, **kwargs):
        self.client.publish(arg1, arg2, ..., repeat=100)
```

**Comment by mtbentley** [comment]
Simply putting a for loop in publish runs into some nasty synchronization errors.

``` python
    def publish(self, topic, payload=None, repeat=1, **kwargs):
        timeout = kwargs.pop('timeout', 5)
        if not hasattr(self, 'mmap'):
            self.mmap = {}
        for i in range(repeat):
            start_time = time.time()
            try:
                err, mid = super(MQTTClient, self).publish(
                    topic,
                    payload=payload,
                    **kwargs
                )
                if err:
                    raise ValueError(err)
                self.mmap[mid] = Message(topic, payload, start_time, timeout)
            except Exception as e:
                total_time = time_delta(start_time, time.time())
                fire_locust_failure(
                    request_type='mqtt',
                    name='publish',
                    response_time=total_time,
                    exception=e,
                )
```

```
[2015-12-06 17:53:32,514] sylvester/ERROR/stderr: Exception in thread Thread-780:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib64/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/matthew/projects/mqtt-locust/venv/lib/python2.7/site-packages/paho/mqtt/client.py", line 2287, in _thread_main
    self.loop_forever()
  File "/home/matthew/projects/mqtt-locust/venv/lib/python2.7/site-packages/paho/mqtt/client.py", line 1261, in loop_forever
    rc = self.loop(timeout, max_packets)
  File "/home/matthew/projects/mqtt-locust/venv/lib/python2.7/site-packages/paho/mqtt/client.py", line 827, in loop
    rc = self.loop_write(max_packets)
  File "/home/matthew/projects/mqtt-locust/venv/lib/python2.7/site-packages/paho/mqtt/client.py", line 1098, in loop_write
    rc = self._packet_write()
  File "/home/matthew/projects/mqtt-locust/venv/lib/python2.7/site-packages/paho/mqtt/client.py", line 1525, in _packet_write
    self.on_publish(self, self._userdata, packet['mid'])
  File "/home/matthew/projects/mqtt-locust/mqtt_locust.py", line 93, in _on_publish
    self.check_for_locust_timeouts(end_time)
  File "/home/matthew/projects/mqtt-locust/mqtt_locust.py", line 96, in check_for_locust_timeouts
    timed_out = [mid for mid, msg in self.mmap.iteritems()
RuntimeError: dictionary changed size during iteration
```

**Comment by ajm188** [comment]
Can you clone the dict before iterating it? `a = dict(self.mmap)`. I think this still forces python to iterate the dict, though.

**Comment by mtbentley** [comment]
That seems to work, but I'm still getting a ton of `TimeoutError`s (with timeout=60)

**Comment by mtbentley** [comment]
I'll try to figure out exactly what's happening later tonight

**Comment by ajm188** [comment]
Closed by b6193f8
